### ISSUE TITLE
Implement functions that allow for quests to be built and played

### ISSUE THREAD

**Issue by philipchenaux** [uchicago-cs/chiventure#646]
In sprint 3, we were able to work create a rough  design document for quests as well as implementing the basic struct and new, init, free functions #466 . Now we plan on implementing more complex functions that allow for quests to firstly be designed and built, and then once the quest has been made allowing for it to be player. For now the functions that we plan to implement are specified on the [wiki](https://github.com/uchicago-cs/chiventure/wiki/RPG-NPC:-Independent-Feature:-Quests-Design-Document), but as we begin implementation we will most likely add more functions as required.

**Comment by philipchenaux** [comment]
Based on the main [wiki](https://github.com/uchicago-cs/chiventure/wiki/RPG-NPC:-Independent-Feature:-Quests-Design-Document) that addresses all the initial functionalities that we will be implementing for quests, we have begun the implementation of the following functions:
```C
/* Adds an achievement to a quest
 * returns SUCCESS OR FAILURE
 */
int add_achievement_to_quest(quest_t *quest, achievement_t *achievement_to_add);

/* Starts a quest
 * returns SUCCESS or FAILURE
 */
int start_quest(quest_t *quest);

/* Checks the status of a quest
 * returns status code for various status in the quest:
 * 0: quest has been completed
 * 1: quest has been started but not completed
 * 2: quest has not been started
 */
int quest_status(quest_t *quest);
```
on the [rpg-npc/quests](https://github.com/uchicago-cs/chiventure/tree/rpg-npc/quests) branch. In addition to writing the implementation and header files of the functions above, we are also in the implementation phase of the relevant tests for the functions above.

**Comment by philipchenaux** [comment]
We have implemented the following additional functions:
```C
/* completes an achievement in a quest
 * 
 * Paramter:
 * - quest: pointer to the quest
 * - item_collected: the item that has been collected
 * - npc_met: the npc that was met
 *
 * Returns:
 * - SUCCESs
 * - FAILURE
 */
int complete_achievement(quest_t *quest, item_t *item_collected, npc_t *npc_met);

/* checks if the quest is completed
 * 
 * Paramter:
 * - pointer to the quest
 *
 * Returns:
 * 0 if quest is incomplete, 1 if it is complete
 */
int is_quest_completed(quest_t *quest);
```
And we are now implementing the testing phase of these functions. We are still unsure, however, on the mechanism of actually how to start a quest, whether that involved talking to an npc or picking up an object.

**Comment by philipchenaux** [comment]
In addition to the functions above, we have also added the following final function that allow for a basic implementation of quests in a game:
```C
/* Rewards the prize to the player once the quest has been completed
 * This also checks for completion, and does nothing if not completed
 *
 * Parameter:
 * - quest: pointer to quest_t
 * 
 * Returns:
 * - an item (possible to be added into the inventory of the player)
 * - If the quest is not completed the function will return a null pointer
 * - The status of the quest should first be checked before this function is called
 */
item_t *quest_completed(quest_t *quest);
```
Now with all the functions, we have  been able to implement individual tests for each of these functions that have been  pushed to  the branch.

**Comment by philipchenaux** [comment]
We now have submitted a pull request #776 , that includes the implementation and testing of all functions mentioned above, and specified on the wiki: https://github.com/uchicago-cs/chiventure/wiki/RPG-NPC:-Independent-Feature:-Quests-Design-Document . This pull request still however, is only a basic implementation of quests that allow the  user to play through a simple variation of a quest. On the wiki we have included additional features that can be implemented in  the future, these include including battles within quests, etc. For now given that we have submitted this pull request, we will close this issue after the pull request is merged with the dev branch and then add further implementation features of quests along with our wiki on the backlog.
